cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll

```diff
@@ -0,0 +1,38 @@
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.DataFlow
+import semmle.code.cpp.models.interfaces.Taint
+
+/**
+ * The standard function `memset` and its assorted variants
+ */
+class MemsetFunction extends ArrayFunction, DataFlowFunction, TaintFunction {
+  MemsetFunction() {
+    this instanceof TopLevelFunction and
+    (
+      this.hasName("memset") or
+      this.hasName("__builtin_memset") or
+      this.hasName("FillMemory")
+    )
+  }
+
+  override predicate hasArrayOutput(int bufParam) { bufParam = 0 }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isInParameter(0) and
+    output.isOutReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (input.isInParameter(1) or input.isInParameter(2)) and
+    (
+      output.isOutParameterPointer(0) or
+      output.isOutReturnPointer()
+    )
```

Comment:
I don't see the value in any of the four flow combinations defined by this predicate. I can't think of a practical query where we'd want such flow.

Reply:
So what do you think is the correct taint model here?  In my way of thinking, both the initializer value and the length affect the output memory buffer.

Reply:
That's technically true, but traditionally we haven't had much luck with taint through low-bandwidth channels like `strlen`. It has been the source of many false positives and was recently disabled from the `security.TaintTracking` library.

I suggest that we don't give a taint model for memset at all unless we have an example of a query and a snapshot where it would be beneficial.